### PR TITLE
initial

### PR DESCRIPTION
**Pull request by NarikNarik**

### CODE REVIEW

File: App.js [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+
+  addTodo = () => {
+    const newTodos = [...this.state.todos, {text: this.state.text}]
+    if (this.state.text !== '') {
+    this.setState({todos: newTodos, text: ''})
+    }
+  }
+
+  deleteTodo = (index) => {
+    const newTodos = [...this.state.todos]
```
Comment (Statix411) [inline App.js]: @NarikNarik why "newTodos"?

Comment (NarikNarik) [reply App.js]: @Statix411 need to rename. I was thinking about it

File: App.js [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+
+  completeTodo = (index) => {
+		const newTodos = [...this.state.todos]
+		newTodos[index].isCompleted = !newTodos[index].isCompleted
+		this.setState({todos: newTodos})
+	}
+
+  editTodo = (index) => {
+    const newTodos = [...this.state.todos]
+    const todo = newTodos[index]
```
Comment (Statix411) [inline App.js]: @NarikNarik here too. 'newTodos'

File: App.js [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+    newTodos[index].isCompleted = !newTodos[index].isCompleted
+    if (this.state.text !== '') {
+    todo.text = this.state.text
```
Comment (Statix411) [inline App.js]: @NarikNarik wrong tabulation

File: App.js [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+    }
+  }
+
+  deleteTodo = (index) => {
+    const newTodos = [...this.state.todos]
+		newTodos.splice(index, 1)
+    this.setState({todos: newTodos})
+  }
+
+  completeTodo = (index) => {
```
Comment (Statix411) [inline App.js]: @NarikNarik change 'completeTodo' to 'toggleStatus'

File: App.js [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+    this.state = {
+      todos: [],
+      text: ''
+    }
+  }
+
+  addTodo = () => {
+    const newTodos = [...this.state.todos, {text: this.state.text}]
+    if (this.state.text !== '') {
+    this.setState({todos: newTodos, text: ''})
```
Comment (Statix411) [inline App.js]: @NarikNarik wrong tabulation

File: App.js [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+        <Text>
+          {this.state.todos.length + ' Todos'}
+          {this.state.todos.filter(todo => !todo.isCompleted).length +
+            'Not completed Todos'}
+        </Text>
+        </View>
+        {
+          this.state.todos && this.state.todos.map((todo, index) => {
+            let textLine = todo.isCompleted ? 'line-through': 'none'
+            return <Todo item={todo} key={index}
```
Comment (Statix411) [inline App.js]: @NarikNarik bad key, you must avoid to use an index of an iteration

Comment (Statix411) [reply App.js]: @NarikNarik you need to generate a unique identifier for each element

File: App.js [hunk 7]
```diff
@@ -0,0 +1,87 @@
+import React from 'react'
+import Todo from './Todo'
+import { StyleSheet, Text, View, TextInput, Button } from 'react-native'
+
+export default class App extends React.Component {
```
Comment (Statix411) [inline App.js]: @NarikNarik change `App` to `Main`

File: App.js [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+    this.setState({todos: newTodos, text: ''})
+    }
+  }
+
+  render() {
+    return (
+      <View style={styles.container}>
+        <View>
+          <TextInput
+            style={{height: 26, borderWidth: 1, backgroundColor: '#fff', borderColor: 'green'}}
```
Comment (Statix411) [inline App.js]: @NarikNarik the styles are hardcoded, use css classes

File: App.js [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+            value={this.state.text}
+          />
+          <Button
+            color='green'
+            onPress={this.addTodo}
+            title='Add Todo'
+          />
+        <Text>
+          {this.state.todos.length + ' Todos'}
+          {this.state.todos.filter(todo => !todo.isCompleted).length +
```
Comment (Statix411) [inline App.js]: @NarikNarik move this code to a separate function

File: App.js [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+            'Not completed Todos'}
+        </Text>
+        </View>
+        {
+          this.state.todos && this.state.todos.map((todo, index) => {
```
Comment (Statix411) [inline App.js]: @NarikNarik you dont need this condition because todos variable is always array

File: App.js [hunk 11]
```diff
@@ -0,0 +1,87 @@
...
+            let textLine = todo.isCompleted ? 'line-through': 'none'
+            return <Todo item={todo} key={index}
+              deleteTodo={this.deleteTodo.bind(this, index)}
```
Comment (Statix411) [inline App.js]: @NarikNarik just `delete`

File: App.js [hunk 12]
```diff
@@ -0,0 +1,87 @@
...
+              completeTodo={this.completeTodo.bind(this, index)}
```
Comment (Statix411) [inline App.js]: @NarikNarik same

File: App.js [hunk 13]
```diff
@@ -0,0 +1,87 @@
...
+          />
+        <Text>
+          {this.state.todos.length + ' Todos'}
+          {this.state.todos.filter(todo => !todo.isCompleted).length +
+            'Not completed Todos'}
+        </Text>
+        </View>
+        {
+          this.state.todos && this.state.todos.map((todo, index) => {
+            let textLine = todo.isCompleted ? 'line-through': 'none'
```
Comment (Statix411) [inline App.js]: @NarikNarik Todo component should be responsible for it

File: Todo.js [hunk 14]
```diff
@@ -0,0 +1,33 @@
+import React from 'react'
+import { StyleSheet, Text, View, Button,TouchableHighlight } from 'react-native'
+
+export default class Todo extends React.Component {
+  constructor(props){
```
Comment (Statix411) [inline Todo.js]: @NarikNarik you should use constructor when you wont to define state, bind functions or something else

File: Todo.js [hunk 15]
```diff
@@ -0,0 +1,33 @@
...
+    super(props)
+  }
+
+  render() {
+    return (
+      <View style={{flexDirection: 'row'}}>
```
Comment (Statix411) [inline Todo.js]: @NarikNarik use css classes

File: Todo.js [hunk 16]
```diff
@@ -0,0 +1,33 @@
...
+        <TouchableHighlight
+           style={{alignItems: 'center', backgroundColor: '#fff',
```
Comment (Statix411) [inline Todo.js]: @NarikNarik use css classes

File: Todo.js [hunk 17]
```diff
@@ -0,0 +1,33 @@
...
+             padding: 10}}
+           onPress={this.props.completeTodo}
+          >
+          <Text style={{textDecorationLine: this.props.textLine}}>{this.props.item.text + ''}</Text>
```
Comment (Statix411) [inline Todo.js]: @NarikNarik use css classes

Comment (Statix411) [reply Todo.js]: @NarikNarik and why do you concatinate the empty string?

**Comment by zheoreh** [comment]
@ByShock please view our review.

**Comment by zheoreh** [comment]
@NarikNarik please do the requested changes.

File: App.js (hunk 1 above)
Comment (NarikNarik) [reply App.js]: @Statix411 Done

File: App.js (hunk 2 above)
Comment (NarikNarik) [reply App.js]: @Statix411 Done

File: App.js (hunk 3 above)
Comment (NarikNarik) [reply App.js]: @Statix411 Done

File: App.js (hunk 4 above)
Comment (NarikNarik) [reply App.js]: @Statix411 Done

File: App.js (hunk 5 above)
Comment (NarikNarik) [reply App.js]: @Statix411 Done

File: App.js (hunk 6 above)
Comment (NarikNarik) [reply App.js]: @Statix411 Done

File: App.js (hunk 7 above)
Comment (NarikNarik) [reply App.js]: @Statix411 Done

File: App.js [hunk 18]
```diff
@@ -0,0 +1,87 @@
...
+export default class App extends React.Component {
+  constructor(props){
+    super(props)
+    this.state = {
+      todos: [],
+      text: ''
+    }
+  }
+
+  addTodo = () => {
```
[the remark this answers is not in the document]
Comment (NarikNarik) [reply App.js]: @Statix411 Done

File: App.js [hunk 19]
```diff
@@ -0,0 +1,87 @@
...
+    const newTodos = [...this.state.todos, {text: this.state.text}]
+    if (this.state.text !== '') {
```
[the remark this answers is not in the document]
Comment (NarikNarik) [reply App.js]: @Statix411 Done

File: App.js [hunk 20]
```diff
@@ -0,0 +1,87 @@
...
+    this.setState({todos: newTodos, text: ''})
+    }
+  }
+
+  deleteTodo = (index) => {
```
[the remark this answers is not in the document]
Comment (NarikNarik) [reply App.js]: @Statix411 Done

File: App.js [hunk 21]
```diff
@@ -0,0 +1,87 @@
...
+    const newTodos = [...this.state.todos]
+		newTodos.splice(index, 1)
+    this.setState({todos: newTodos})
+  }
+
+  completeTodo = (index) => {
+		const newTodos = [...this.state.todos]
```
[the remark this answers is not in the document]
Comment (NarikNarik) [reply App.js]: @Statix411 Done

File: App.js [hunk 22]
```diff
@@ -0,0 +1,87 @@
...
+		newTodos[index].isCompleted = !newTodos[index].isCompleted
+		this.setState({todos: newTodos})
+	}
+
+  editTodo = (index) => {
+    const newTodos = [...this.state.todos]
```
[the remark this answers is not in the document]
Comment (NarikNarik) [reply App.js]: @Statix411 Done

File: App.js [hunk 23]
```diff
@@ -0,0 +1,87 @@
...
+    const todo = newTodos[index]
+    newTodos[index].isCompleted = !newTodos[index].isCompleted
+    if (this.state.text !== '') {
```
[the remark this answers is not in the document]
Comment (NarikNarik) [reply App.js]: @Statix411 Done

File: App.js (hunk 8 above)
Comment (NarikNarik) [reply App.js]: @Statix411 Done

File: App.js (hunk 9 above)
Comment (NarikNarik) [reply App.js]: @Statix411 Done

File: App.js (hunk 10 above)
Comment (NarikNarik) [reply App.js]: @Statix411 Done

File: App.js (hunk 11 above)
Comment (NarikNarik) [reply App.js]: @Statix411 Done

File: App.js (hunk 12 above)
Comment (NarikNarik) [reply App.js]: @Statix411 Done

File: App.js (hunk 13 above)
Comment (NarikNarik) [reply App.js]: @Statix411 Done

File: Todo.js (hunk 14 above)
Comment (NarikNarik) [reply Todo.js]: @Statix411 Done

File: Todo.js (hunk 15 above)
Comment (NarikNarik) [reply Todo.js]: @Statix411 Done

File: Todo.js (hunk 16 above)
Comment (NarikNarik) [reply Todo.js]: @Statix411 Done

File: Todo.js (hunk 17 above)
Comment (NarikNarik) [reply Todo.js]: @Statix411 Done

File: components/Main.js [hunk 24]
```diff
@@ -0,0 +1,140 @@
...
+  getCompletedTodosCount = () => {
+    return this.state.todos.filter(todo => !todo.isCompleted).length
+  }
+
+  render() {
+    return (
+      <View style={styles.container}>
+        <View>
+        <View>
+          <Text style={styles.counterAll}>All: {this.getTotalTodosCount()}</Text>
```
Comment (Statix411) [inline components/Main.js]: @NarikNarik I think here function is not necessary

Comment (Statix411) [inline components/Main.js]: @NarikNarik I think here function is not necessary

Comment (zheoreh) [reply components/Main.js]: @Statix411 functions are not superfluous)

Comment (Statix411) [reply components/Main.js]: @zheoreh okey)

File: components/Main.js [hunk 25]
```diff
@@ -0,0 +1,140 @@
+import React from 'react'
+import Todo from './Todo'
+import { StyleSheet, Text, View, TextInput, Button } from 'react-native'
+
+export default class App extends React.Component {
```
Comment (zheoreh) [inline components/Main.js]: @NarikNarik `Main`. But rename it to `Todos`. Components must have concrete name. We work with todos here so it should be names `Todos`

File: components/Main.js [hunk 26]
```diff
@@ -0,0 +1,140 @@
...
+
+  constructor(props){
+    super(props)
+
+    this.state = {
+      todos: [],
```
Comment (zheoreh) [inline components/Main.js]: @NarikNarik `items`

File: components/Main.js [hunk 27]
```diff
@@ -0,0 +1,140 @@
...
+      text: '',
+      id: 0
+    }
+  }
+
+  add = () => {
```
Comment (zheoreh) [inline components/Main.js]: @NarikNarik let it be addItem

File: components/Main.js [hunk 28]
```diff
@@ -0,0 +1,140 @@
...
+    let {id, text} = this.state
+
+    let item = {
+      text: text,
+      id: id++,
+      completed: true
+     }
+
+    if (!text) return
```
Comment (zheoreh) [inline components/Main.js]: @NarikNarik move higher and trim a text

File: components/Main.js [hunk 29]
```diff
@@ -0,0 +1,140 @@
...
+
+    this.setState((state) => ({
+      todos: [...state.todos, item],
+      text: '',
+      id
+    }))
+  }
+
+  edit = (id) => {
```
Comment (zheoreh) [inline components/Main.js]: @NarikNarik `editItem`

File: components/Main.js [hunk 30]
```diff
@@ -0,0 +1,140 @@
...
+    let {text} = this.state
+
+    if (!text) return
+
+    this.setState((state) => ({
+      todos: state.todos.map((item) => {
+        if (item.id === id)
+          return {...item, text: text, isCompleted: false}
```
Comment (zheoreh) [inline components/Main.js]: @NarikNarik I'm not sure that we should change the status. I would say that we can edit only not completed task but it's for an option.

File: components/Main.js [hunk 31]
```diff
@@ -0,0 +1,140 @@
...
+
+        return item
+      }),
+      text: ''
+    }))
+  }
+
+  toggleStatus = (id) => {
```
Comment (zheoreh) [inline components/Main.js]: @NarikNarik `toggleItemStatus`

File: components/Main.js [hunk 32]
```diff
@@ -0,0 +1,140 @@
...
+    this.setState((state) => ({
+      todos: state.todos.map((item) => {
+        if (item.id === id)
+          return {...item, isCompleted: !item.isCompleted}
```
Comment (zheoreh) [inline components/Main.js]: @NarikNarik wrong field name. Correct is `completed`

File: components/Main.js [hunk 33]
```diff
@@ -0,0 +1,140 @@
...
+
+        return item
+      })
+    }))
+  }
+
+  delete = (id) => {
```
Comment (zheoreh) [inline components/Main.js]: @NarikNarik `deleteItem`

File: components/Main.js [hunk 34]
```diff
@@ -0,0 +1,140 @@
...
+    this.setState((state) => ({
+      todos: state.todos.filter((item) => {
+        return item.id !==id
```
Comment (zheoreh) [inline components/Main.js]: @NarikNarik `!== id`

File: components/Main.js [hunk 35]
```diff
@@ -0,0 +1,140 @@
...
+      })
+    }))
+  }
+
+  getTotalTodosCount = () => {
```
Comment (zheoreh) [inline components/Main.js]: @NarikNarik `getTotalItemsCount` or `geItemsCount`

File: components/Main.js [hunk 36]
```diff
@@ -0,0 +1,140 @@
...
+    return this.state.todos.length
+  }
+
+  getCompletedTodosCount = () => {
```
Comment (zheoreh) [inline components/Main.js]: @NarikNarik same rule...

File: components/Main.js [hunk 37]
```diff
@@ -0,0 +1,140 @@
...
+              onPress={this.add}
+              title='Add Todo'
+            />
+          </View>
+        </View>
+        {
+          this.state.todos.map((todo) => {
+            let textLine = todo.isCompleted
+            return <Todo item={todo} key={todo.id}
+              deleteTodo={this.delete.bind(this, todo.id)}
```
Comment (zheoreh) [inline components/Main.js]: @NarikNarik `deleteTodo` -> `delete`

File: components/Main.js [hunk 38]
```diff
@@ -0,0 +1,140 @@
...
+            <Button
+              color='green'
+              onPress={this.add}
+              title='Add Todo'
+            />
+          </View>
+        </View>
+        {
+          this.state.todos.map((todo) => {
+            let textLine = todo.isCompleted
```
Comment (zheoreh) [inline components/Main.js]: @NarikNarik remove. Todo component must be responsible for it. And don't forget your variable is `completed`

File: components/Main.js [hunk 39]
```diff
@@ -0,0 +1,140 @@
...
+            return <Todo item={todo} key={todo.id}
+              deleteTodo={this.delete.bind(this, todo.id)}
+              toggleStatus={this.toggleStatus.bind(this, todo.id)}
+              textLine={textLine}
+              editTodo={this.edit.bind(this, todo.id)}
```
Comment (zheoreh) [inline components/Main.js]: @NarikNarik `editTodo` -> `edit`

**Comment by zheoreh** [comment]
@NarikNarik please do the changes.

File: components/Main.js (hunk 39 above)
Comment (NarikNarik) [reply components/Main.js]: @zheoreh Done

File: components/Main.js (hunk 38 above)
Comment (NarikNarik) [reply components/Main.js]: @zheoreh Done

File: components/Main.js (hunk 37 above)
Comment (NarikNarik) [reply components/Main.js]: @zheoreh Done

File: components/Main.js (hunk 36 above)
Comment (NarikNarik) [reply components/Main.js]: @zheoreh Done

File: components/Main.js (hunk 35 above)
Comment (NarikNarik) [reply components/Main.js]: @zheoreh Done

File: components/Main.js (hunk 34 above)
Comment (NarikNarik) [reply components/Main.js]: @zheoreh Done

File: components/Main.js (hunk 33 above)
Comment (NarikNarik) [reply components/Main.js]: @zheoreh Done

File: components/Main.js (hunk 32 above)
Comment (NarikNarik) [reply components/Main.js]: @zheoreh Done

File: components/Main.js (hunk 31 above)
Comment (NarikNarik) [reply components/Main.js]: @zheoreh Done

File: components/Main.js (hunk 30 above)
Comment (NarikNarik) [reply components/Main.js]: @zheoreh Done

File: components/Main.js (hunk 29 above)
Comment (NarikNarik) [reply components/Main.js]: @zheoreh Done

File: components/Main.js (hunk 28 above)
Comment (NarikNarik) [reply components/Main.js]: @zheoreh Done

File: components/Main.js (hunk 27 above)
Comment (NarikNarik) [reply components/Main.js]: @zheoreh Done

File: components/Main.js (hunk 26 above)
Comment (NarikNarik) [reply components/Main.js]: @zheoreh Done

File: components/Main.js (hunk 25 above)
Comment (NarikNarik) [reply components/Main.js]: @zheoreh Done

File: components/Todos.js [hunk 40]
```diff
@@ -8,80 +8,84 @@ export default class App extends React.Component {
...
-      id: 0
+      id: 0,
+      completed: false
     }
   }
 
-  add = () => {
-    let {id, text} = this.state
+  addItem = () => {
+    let {id, text, completed} = this.state
```
Comment (zheoreh) [inline components/Todos.js]: @NarikNarik why have you added completed?

File: components/Todos.js [hunk 41]
```diff
@@ -8,80 +8,84 @@ export default class App extends React.Component {
...
 
+    text = text.trim()
     let item = {
-      text: text,
-      id: id++,
-      completed: true
+      text,
```
Comment (zheoreh) [inline components/Todos.js]: @NarikNarik a text still isn't trimmed

File: components/Todos.js [hunk 42]
```diff
@@ -8,80 +8,84 @@ export default class App extends React.Component {
...
+      id,
+      completed
      }
-
-    if (!text) return
+     if (!text) return
```
Comment (zheoreh) [inline components/Todos.js]: @NarikNarik this condition hasn't been moved higher. Have a look at 40th line.

File: components/Todos.js [hunk 43]
```diff
@@ -8,80 +8,84 @@ export default class App extends React.Component {
...
+      id: ++id,
+      completed: false
     }))
   }
 
-  edit = (id) => {
+  editItem = (id) => {
     let {text} = this.state
-
+    text = text.trim()
```
Comment (zheoreh) [inline components/Todos.js]: @NarikNarik same should have been done for addItem too.

File: components/Todos.js [hunk 44]
```diff
@@ -8,80 +8,84 @@ export default class App extends React.Component {
...
     this.setState((state) => ({
-      todos: state.todos.map((item) => {
-        if (item.id === id)
-          return {...item, text: text, isCompleted: false}
-
+      items: state.items.map((item) => {
+        if (item.id === id && !item.completed) {
+          return {...item, text}
+        } else if (item.id === id && item.completed) {
+          alert("can't edit")
```
Comment (zheoreh) [inline components/Todos.js]: @NarikNarik дич. ты вообще это тестил?)

File: components/Todos.js [hunk 45]
```diff
@@ -8,80 +8,84 @@ export default class App extends React.Component {
...
-    if (!text) return
+     if (!text) return
 
     this.setState((state) => ({
-      todos: [...state.todos, item],
+      items: [...state.items, item],
       text: '',
-      id
+      id: ++id,
+      completed: false
```
Comment (zheoreh) [inline components/Todos.js]: @NarikNarik why?

File: components/Todos.js [hunk 46]
```diff
@@ -91,19 +95,19 @@ export default class App extends React.Component {
...
               title='Add Todo'
             />
           </View>
         </View>
         {
-          this.state.todos.map((todo) => {
-            let textLine = todo.isCompleted
+          this.state.items.map((todo) => {
+
             return <Todo item={todo} key={todo.id}
```
Comment (zheoreh) [inline components/Todos.js]: @NarikNarik please rename item prop to data. Todo item sound bad, doesn't it?

File: components/Todos.js [hunk 47]
```diff
@@ -91,19 +95,19 @@ export default class App extends React.Component {
...
-              deleteTodo={this.delete.bind(this, todo.id)}
-              toggleStatus={this.toggleStatus.bind(this, todo.id)}
-              textLine={textLine}
-              editTodo={this.edit.bind(this, todo.id)}
+              delete={this.deleteItem.bind(this, todo.id)}
+              toggleItemStatus={this.toggleItemStatus.bind(this, todo.id)}
+              textLine={todo.completed}
```
Comment (zheoreh) [inline components/Todos.js]: @NarikNarik what did I say in previous review?

File: components/Todos.js (hunk 42 above)
Comment (NarikNarik) [reply components/Todos.js]: @zheoreh Done

File: components/Todos.js (hunk 44 above)
Comment (NarikNarik) [reply components/Todos.js]: @zheoreh This "дич" worked. but I did ok

File: components/Todos.js (hunk 45 above)
Comment (NarikNarik) [reply components/Todos.js]: @zheoreh Done

File: components/Todos.js (hunk 46 above)
Comment (NarikNarik) [reply components/Todos.js]: @zheoreh Done

File: components/Todos.js (hunk 47 above)
Comment (NarikNarik) [reply components/Todos.js]: @zheoreh Done)

File: components/Todos.js (hunk 40 above)
Comment (NarikNarik) [reply components/Todos.js]: @zheoreh Done

File: components/Todos.js (hunk 43 above)
Comment (NarikNarik) [reply components/Todos.js]: @zheoreh There also

File: components/Todos.js (hunk 44 above)
Comment (zheoreh) [reply components/Todos.js]: @NarikNarik а ну да, у тебя только раз алерт отработает, согласен, но это не место для этого.

File: components/Todos.js (hunk 43 above)
Comment (zheoreh) [reply components/Todos.js]: @NarikNarik I haven't seen. I will check

File: components/Todos.js [hunk 48]
```diff
@@ -16,37 +16,35 @@ export default class App extends React.Component {
...
     this.setState((state) => ({
       items: state.items.map((item) => {
-        if (item.id === id && !item.completed) {
+        if(item.id !== id){
+          return item
+        } else if (!item.completed) {
           return {...item, text}
-        } else if (item.id === id && item.completed) {
+        } else {
           alert("can't edit")
```
Comment (zheoreh) [inline components/Todos.js]: @NarikNarik please implement a check before modifying the state. You must show an alert at once before any changes. `setState` is responsible only for data.

File: components/Todos.js [hunk 49]
```diff
@@ -101,13 +99,12 @@ export default class App extends React.Component {
...
-            return <Todo item={todo} key={todo.id}
-              delete={this.deleteItem.bind(this, todo.id)}
-              toggleItemStatus={this.toggleItemStatus.bind(this, todo.id)}
-              textLine={todo.completed}
-              edit={this.editItem.bind(this, todo.id)}
+          this.state.items.map((item) => {
+            return <Todo data={item} key={item.id}
+              delete={this.deleteItem.bind(this, item.id)}
+              toggleItemStatus={this.toggleItemStatus.bind(this, item.id)}
+              completed={item.completed}
```
Comment (zheoreh) [inline components/Todos.js]: @NarikNarik you have a data inside of Todo so why do you pass the field as a prop? Third time, dude...

File: components/Todo.js [hunk 50]
```diff
@@ -5,12 +5,12 @@ export default class Todo extends React.Component {
...
     return (
-      <View style={styles.item}>
+      <View style={styles.root}>
         <TouchableHighlight
-           style={styles.itemList}
+           style={styles.list}
            onPress={this.props.toggleItemStatus}
           >
-          <Text style={(this.props.textLine)? styles.todoTextLine : styles.todoText}>{this.props.item.text}</Text>
+          <Text style={(this.props.completed)? styles.TextLine : styles.Text}>{this.props.data.text}</Text>
```
[the remark this answers is not in the document]
Comment (NarikNarik) [reply components/Todo.js]: @zheoreh done

File: components/Todos.js (hunk 48 above)
Comment (NarikNarik) [reply components/Todos.js]: @zheoreh done

File: components/Todos.js (hunk 49 above)
Comment (NarikNarik) [reply components/Todos.js]: @zheoreh done

File: components/Todos.js [hunk 51]
```diff
@@ -35,17 +35,18 @@ export default class App extends React.Component {
     let {text} = this.state
     text = text.trim()
     if (!text) return
+    let items = this.state.items.map((item) => {
+      if (item.id !== id) {
+        return item
+      } else if (!item.completed) {
+        return {...item, text: text}
+      } else {
+        alert("Can't edit")
```
Comment (zheoreh) [inline components/Todos.js]: @Statix411 no. You should use map only to update specified todo. As I said before you must do a check before any update. Any portion of code must do one thing. Разделяй и властвуй. Ты сперва по айди должен достать нужную тудушку, сделать проверку, только потом модифицировать данные. Такой код будет называться поддерживаемым, потому что проверка вынесена, ее можно заменить не вникая в этот map ну или вообще можно ее легко убрать.

Comment (NarikNarik) [reply components/Todos.js]: @zheoreh change

File: components/Todos.js [hunk 52]
```diff
@@ -32,21 +31,21 @@ export default class App extends React.Component {
   }
 
   editItem = (id) => {
+    let items = [...this.state.items]
```
Comment (zheoreh) [inline components/Todos.js]: @NarikNarik there is no need to copy items because you aren't going to modify them.

File: components/Todos.js [hunk 53]
```diff
@@ -32,21 +31,21 @@ export default class App extends React.Component {
...
+    let item = items.find((item) => {
+      return item.id === id
+    })
+
+    if (item.completed) return alert('dwqd')
```
Comment (zheoreh) [inline components/Todos.js]: @NarikNarik `dwqd`? =DDD

File: components/Todos.js [hunk 54]
```diff
@@ -36,7 +36,7 @@ export default class App extends React.Component {
       return item.id === id
     })
 
-    if (item.completed) return alert('dwqd')
+    if (item.completed) return alert("Ca'nt Edit")
```
Comment (zheoreh) [inline components/Todos.js]: @NarikNarik ну почти) ты там что на поезд опаздываешь?

File: components/Todos.js (hunk 52 above)
Comment (NarikNarik) [reply components/Todos.js]: @zheoreh done

File: components/Todos.js (hunk 53 above)
Comment (NarikNarik) [reply components/Todos.js]: @zheoreh LoL)

File: components/Todos.js (hunk 54 above)
Comment (NarikNarik) [reply components/Todos.js]: @zheoreh ))

**Pull request merged**